pkg/tessera/tessera.go

```diff
@@ -0,0 +1,139 @@
+	noteSigner, err := note.NewNoteSigner(ctx, origin, signer)
+	if err != nil {
+		return nil, fmt.Errorf("getting note signer: %w", err)
+	}
+	appender, reader, err := tessera.NewAppender(ctx, driver,
+		tessera.NewAppendOptions().WithCheckpointSigner(noteSigner),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting tessera appender: %w", err)
+	}
+	awaiter := tessera.NewIntegrationAwaiter(ctx, reader.ReadCheckpoint, 1*time.Second)
+	return &Storage{
+		origin:     origin,
+		awaiter:    awaiter,
+		addFn:      appender.Add,
+		readTileFn: reader.ReadTile,
+	}, nil
+}
+
+// Add adds a Tessera entry to the log, waits for it to be sequenced into the log,
+// and returns the log index and inclusion proof as a TransparencyLogEntry object.
+func (s *Storage) Add(ctx context.Context, entry *tessera.Entry) (*rekor_pb.TransparencyLogEntry, error) {
+	idx, checkpointBody, err := s.addEntry(ctx, entry)
+	if err != nil {
+		return nil, fmt.Errorf("add entry: %w", err)
+	}
+	inclusionProof, err := s.buildProof(ctx, idx, checkpointBody)
+	if err != nil {
+		return nil, fmt.Errorf("building inclusion proof: %w", err)
+	}
+	return &rekor_pb.TransparencyLogEntry{
+		LogIndex:       idx.i,
+		InclusionProof: inclusionProof,
+	}, nil
+}
+
+// ReadTile looks up the tile at the given level, index within the level, and
+// width of the tile if partial, and returns the raw bytes of the tile.
+func (s *Storage) ReadTile(ctx context.Context, level, index uint64, p uint8) ([]byte, error) {
+	tile, err := s.readTileFn(ctx, level, index, p)
+	if err != nil {
+		return nil, fmt.Errorf("reading tile level %d index %d p %d: %w", level, index, p, err)
+	}
+	return tile, nil
+}
+
+func (s *Storage) addEntry(ctx context.Context, entry *tessera.Entry) (*safeInt64, []byte, error) {
+	idx, checkpointBody, err := s.awaiter.Await(ctx, s.addFn(ctx, entry))
+	if err != nil {
+		return nil, nil, fmt.Errorf("await: %w", err)
+	}
+	safeIdx, err := newSafeInt64(idx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid index: %w", err)
+	}
+	return safeIdx, checkpointBody, nil
+}
+
+func (s *Storage) buildProof(ctx context.Context, idx *safeInt64, signedCheckpoint []byte) (*rekor_pb.InclusionProof, error) {
+	checkpoint, err := unmarshalCheckpoint(signedCheckpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling checkpoint: %w", err)
+	}
+	proofBuilder, err := client.NewProofBuilder(ctx, checkpoint, s.ReadTile)
+	if err != nil {
+		return nil, fmt.Errorf("new proof builder: %w", err)
+	}
+	proof, err := proofBuilder.InclusionProof(ctx, idx.u)
+	if err != nil {
+		return nil, fmt.Errorf("generating inclusion proof: %w", err)
+	}
+	hashes := make([][]byte, len(proof))
+	for i, h := range proof {
+		dst := make([]byte, hex.EncodedLen(len(h)))
+		hex.Encode(dst, h)
+		hashes[i] = dst
+	}
+	safeCheckpointSize, err := newSafeInt64(checkpoint.Size)
+	if err != nil {
+		return nil, fmt.Errorf("invalid tree size: %d", checkpoint.Size)
+	}
```

Comment:
What do you think about verifying the proof before returning it? I think it's a good practice just to double check something wasn't corrupted. We can later set up a metric for it and alert if it is ever incremented.

```
if err := proof.VerifyInclusion(rfc6962.DefaultHasher, idx.i, safeCheckpointSize, entry.LeafHash(), proof, checkpoint.Hash); err != nil { ... }
```

Reply:
I'm a little worried about the extra overhead this adds but I can add it for now and we can see if it becomes a problem later.